dev/src/watch.ts

```diff
@@ -687,6 +540,113 @@ abstract class Watch {
       });
   }
 
+  /**
+   * Handles 'data' events and closes the stream if the response type is
+   * invalid.
+   * @private
+   */
+  private onData(proto: api.IListenResponse): void {
+    if (proto.targetChange) {
+      logger('Watch.onSnapshot', this.requestTag, 'Processing target change');
+      const change = proto.targetChange;
+      const noTargetIds = !change.targetIds || change.targetIds.length === 0;
+      if (change.targetChangeType === 'NO_CHANGE') {
+        if (noTargetIds && change.readTime && this.current) {
+          // This means everything is up-to-date, so emit the current
+          // set of docs as a snapshot, if there were changes.
+          this.pushSnapshot(
+            Timestamp.fromProto(change.readTime),
+            change.resumeToken!
+          );
+        }
+      } else if (change.targetChangeType === 'ADD') {
+        if (WATCH_TARGET_ID !== change.targetIds![0]) {
+          this.closeStream(Error('Unexpected target ID sent by server'));
+        }
+      } else if (change.targetChangeType === 'REMOVE') {
+        let code = 13;
+        let message = 'internal error';
+        if (change.cause) {
+          code = change.cause.code!;
+          message = change.cause.message!;
+        }
+        // @todo: Surface a .code property on the exception.
```

Comment:
I think we may have now earned a Moma badge by moving this TODO around more than 10 times :)